asahi_firmware/multitouch.py

```diff
@@ -29,7 +29,59 @@ def unmunge(el, idmap):
 
     return plistlib.loads(ET.tostring(pl))
 
-def plist_to_bin(plist):
+LOAD_COMMAND_INIT_PAYLOAD = 0
+LOAD_COMMAND_SEND_BLOB = 1
+LOAD_COMMAND_SEND_PROPERTY = 2
+
+def plist_to_bin_touchscreen(plist):
+    def swap32(i):
+        a = (i >> 8) & 0xFF
+        b = i & 0xFF
+        c = (i >> 24) & 0xFF
+        d = (i >> 16) & 0xFF
+        return a << 24 | b << 16 | c << 8 | d
+    def serialize(plist):
+        for i in plist:
+            if i["Type"] == "Config":
+                init = i["Config"]["SPI Config"]["Init Payload"]
+                yield struct.pack("<II", LOAD_COMMAND_INIT_PAYLOAD, len(init))
+                yield init
+                realign = len(init) % 4
+                if realign != 0:
+                    yield bytes([0] * (4 - realign))
+        for i in plist:
+            pkt_len = 0
+            if i['Type'] == 'Binary':
+                payload = i['Payload']
+                pkt_len = 14 + len(payload)
+                yield struct.pack("<II", LOAD_COMMAND_SEND_BLOB, pkt_len)
+                addr = i['Address']
+                hdr = struct.pack(">HHI", 0x3001, (len(payload) + 3) // 4, swap32(addr))
+                yield hdr
+                yield struct.pack(">H", sum(hdr[2:]))
+                for i in range(len(payload) // 4):
+                    yield bytes([payload[i * 4 + 1], payload[i * 4], payload[i * 4 + 3], payload[i * 4 + 2]])
+                checksum = sum(payload)
+                yield struct.pack(">I", swap32(checksum))
+            elif i['Type'] == 'Property':
+                yield struct.pack("<II", LOAD_COMMAND_SEND_PROPERTY, i['Address'])
```

Comment:
Is this specifically the cal blob, or something more generic? (AFAICT the kernel interprets it specifically as the cal blob)

Reply:
It is _the_ cal blob, but could be extended to be something more generic

Reply:
Then probably best to call the command "LOAD_COMMAND_CAL_BLOB" or similar, since there is no room for expansion (no type argument/etc). We can always add another command later.